nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -287,6 +288,308 @@ fn pick_sled_reservation_target(
+                        .rendezvous_local_storage_dataset_id,
+                    pool_id: zpool_get_result.pool.id(),
+                    sled_id: zpool_get_result.pool.sled_id(),
+                })
+                .collect();
+
+            if candidate_datasets.is_empty() {
+                // if there's no local storage datasets on this sled for this
+                // request's size, then try another sled.
+                info!(
+                    &log,
+                    "sled {sled_target} does not have any candidate datasets \
+                    with available space to satisfy local storage allocation";
+                    "request" => ?request,
+                );
+                return Err(());
+            }
+
+            allocations_to_perform.push(PossibleAllocationsForRequest {
+                request,
+                candidate_datasets,
+            });
+        }
+
+        // From the list of allocations to perform, and all the candidate local
+        // storage datasets that could fit those allocations, find a list of all
+        // valid request -> zpool mappings.
+        //
+        // Start from no allocations made yet, a list of allocations to perform
+        // (stored in `requests`), and all of the available zpools on the sled.
+        // Use a BinaryHeap to get a sorted list.
+
+        let mut queue = BinaryHeap::new();
+
+        queue.push(PossibleAllocations {
+            allocations: vec![],
+            candidates_left: zpools_for_sled
+                .iter()
+                .map(|zpool_get_result| CandidateDataset {
+                    rendezvous_local_storage_dataset_id: zpool_get_result
+                        .rendezvous_local_storage_dataset_id,
+                    pool_id: zpool_get_result.pool.id(),
+                    sled_id: zpool_get_result.pool.sled_id(),
+                })
+                .collect(),
+            request_index: 0,
+        });
+
+        Ok(Self { log: log.clone(), allocations_to_perform, queue })
+    }
+}
+
+impl<'a> Iterator for ValidLocalStorageAllocations<'a> {
+    type Item = NonEmpty<LocalStorageAllocation>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Find each valid allocation by, for each request:
+        //
+        // - selecting a local storage dataset from the list of local storage
+        //   datasets that have not been used yet that could fulfill that
+        //   request
+        //
+        // - removing that selected local storage dataset from the list of
+        //   candidates
+        //
+        // - considering a set of allocations as "valid" if all requests have
+        //   been matched with a local storage dataset.
+
+        while let Some(possible_allocation) = self.queue.pop() {
+            let PossibleAllocations {
+                allocations,
+                candidates_left,
+                request_index,
+            } = possible_allocation;
```

Comment:
feel free to ignore me but, this could also now be 
```suggestion

        loop {
            let PossibleAllocations {
                allocations,
                candidates_left,
                request_index,
            } =  self.queue.pop()?;
```
which...is a different way of saying the same thing that is probably neither worse nor better in any kind of way